rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,406 @@
+Some of the more restrictive security policies are starting to reject RSA2048
+(e.g. [RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Algorithms
+
+These algorithms are being considered for support:
+
+#### RSA
+
+Private key sizes: 2048, 3072, 4096
+
+Signature algorithms: PKCS#1 v1.5, (maybe PSS for TLS and JWT? SSH does not support it)
+
+Digest/hash algorithms: SHA512 for SSH, SHA256 for TLS
+
+Considerations:
+
+* RSA2048 is the current default and deviating from it by default may break
+  compatibility with third-party components and protocols
+* RSA has the most widespread support among all protocols
+* Certain database protocols only support RSA client certs
+  * <https://docs.snowflake.com/en/user-guide/key-pair-auth#step-2-generate-a-public-key>
+* If we must continue to support RSA, we might as well support larger key sizes,
+  3072 and 4096-bit are the most commonly used and supported by e.g. GCP KMS.
+* golang.org/x/crypto/ssh always uses SHA512 hash with RSA public keys
+  <https://github.com/golang/crypto/blob/0ff60057bbafb685e9f9a97af5261f484f8283d1/ssh/certs.go#L443-L445>
+* crypto/x509 always uses SHA256 hash with RSA public keys
+  <https://github.com/golang/go/blob/dbf9bf2c39116f1330002ebba8f8870b96645d87/src/crypto/x509/x509.go#L1411-L1414>
+* ssh only supports the PKCS#1 v1.5 signature scheme
+  <https://datatracker.ietf.org/doc/html/rfc8332>
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+
+#### ECDSA
+
+Curves: P-256, (maybe P-384, P-521?)
+
+Digest/hash algorithms: SHA256, (or SHA384, SHA512 for P-384, P-521)
+
+Considerations:
+
+* ECDSA has good support across SSH and TLS protocols for both client and CA
+  certs.
+* ECDSA key generation is *much* faster than RSA key generation.
+* ECDSA signatures are faster than RSA signatures.
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+
+#### EdDSA
+
+Curves: Ed25519 (this is the only curve supported in Go)
+
+Digest/hash algorithms: none (the full message is signed without hashing)
+
+Considerations:
+
+* There is widespread support for Ed25519 SSH certs.
+* Go libraries support Ed25519 for TLS
+* Support for Ed25519 is *not* widespread in the TLS ecosystem.
+* YubiHSM and GCP KMS do *not* support Ed25519 keys.
+* Ed25519 is considered by some to be the fastest, most secure, most modern
+  option for SSH certs.
+* Ed25519 key generation is *much* faster than RSA key generation.
+* Ed25519 signatures are faster than RSA signatures.
+* FIPS 186-5 approves Ed25519
+* BoringCrypto does not support Ed25519
+  <https://go.googlesource.com/go/+/dev.boringcrypto/src/crypto/tls/boring.go#80>
+
+#### Summary
+
+* We are probably forced to continue unconditionally using RSA for database
+  certs, I'm assuming this would apply to both client and CA.
+* Ed25519 is a modern favourite for SSH, but TLS (and HSM, KMS) support is lacking.
```

Comment:
I would really like to see Ed25519 support for SSH. SSH is our bread and butter and we aim to be the most modern and secure way to do it.

Reply:
In that case I think we have two options:

1. Split user and host SSH/TLS keys so SSH can use Ed25519 and TLS can use (probably) ECDSA. This will require changes to a bunch of APIs, and the way user certs are stored on disk in `~/.tsh`, and probably some changes to they way hosts store their keys/certs, but it should be doable with some work.
2. Also use Ed25519 for TLS certs. The wider TLS ecosystem has very spotty support for Ed25519, but I've tried it out within Teleport and it seems to work just fine. Do client TLS certs ever hit external systems, or do external systems ever need to trust the Teleport CA? I'm not 100% sure but I don't think they do to be honest, for app access it's just JWTs that hit external apps. So we might be able to get away with this.

Reply:
Re: using Ed25519 for TLS certs, third party clients do need to load the certs when using app access with `tsh app login`, which gives you an example curl command to run like `curl   --cert /Users/nklaassen/.tsh/keys/one/nic-app/cluster-one/dumper-x509.pem   --key /Users/nklaassen/.tsh/keys/one/nic   https://dumper.one:4080`. I just tried this out with the default version of `curl` installed on my macbook (v7.88.1) and an Ed25519 client certificate, and curl complained about being unable to load the cert
```
curl: (58) could not load PEM client certificate from /Users/nklaassen/.tsh/keys/one/nic-app/cluster-one/dumper-x509.pem, LibreSSL error error:06FFF09C:digital envelope routines:CRYPTO_internal:unsupported algorithm, (no key found, wrong pass phrase, or wrong file f
```

However, curl 8.1.2 installed through Homebrew is able to load the Ed25519 cert just fine and the request succeeds. The other workaround is to use `tsh proxy app <app>` so that `tsh` is the one loading the client cert and that also works fine with Ed25519 certs.

Reply:
Chatted a bit about this with @jentfoo, I think we agree that long term it would be best to split the SSH/TLS keys so that we can use different algorithms for each. This will allow us to use Ed25519 for SSH keys and ECDSA for TLS keys now, and will give us more options in the future to evolve each of them independently.